MCprep_addon/spawner/mcmodel.py

```diff
@@ -179,33 +175,142 @@ def add_get_material(
+	# This handles cases where a face references a key not explicitly defined.
+	return key
 
+def locate_image(
+	context: Context, textures: Dict[str, str], img: str, model_filepath: Path) -> str:
+	"""
+	Finds and returns the final texture path from a texture key/reference in the model JSON.
+	
+	The search hierarchy is:
+	1. Recursively resolving texture references (e.g., '#side' -> '#all').
+	2. Cleaning the resulting path fragment (e.g., namespace, extension handling).
+	3. Searching the file in this order: Model's Origin Pack (A), Active Scene Resource Pack (B), and Addon Default Resource Pack (C).
+	"""
+
+	# 1. Resolve to final texture key. (e.g. #texture -> #all -> block/texture)
+	# Uses the `get_final_texture_key` function to get the final key
+	final_key = get_final_texture_key(f"#{img}", textures)
+	if final_key is None or final_key not in textures:
+		# Fallback if key is not found or is part of a circular reference
+		local_path = textures.get(img, img)
+	else:
+		local_path = textures[final_key]
+		
+	# Path Cleaning/Normalization
+	# Remove leading path separators AND the .png extension if present. For cleaner & safer searching
+	local_path_clean = local_path.replace("/", os.sep).replace("\\", os.sep)
+	local_path_clean = local_path_clean.lstrip(os.sep) # Prevent leading slash vulnerability
+	if local_path_clean.lower().endswith(".png"): # Prevent '.png.png' double extension
+		local_path_clean = local_path_clean[:-4]
+	
+	# Append the extension once for a complete/clean path fragment
+	texture_path_fragment = local_path_clean + ".png"
+
+	# 2. Local relative paths (starting with `.`)
+	if local_path.startswith('.'):
+		# Path is relative to the model JSON file's directory.
+		directory = os.path.dirname(model_filepath)
+		return os.path.normpath(os.path.join(directory, texture_path_fragment))
+
+	# 3. Parse Namespace (e.g., 'minecraft:block/foo' -> 'block/foo')
+	if ":" in local_path:
+		namespace, local_path_remainder = local_path.split(":", 1)
+	else:
+		namespace = "minecraft"
+		local_path_remainder = local_path
+	
+	# Re-apply path cleaning after namespace removal (safe, catches stray separators)
+	local_path_clean = local_path_remainder.replace("/", os.sep).replace("\\", os.sep)
+	local_path_clean = local_path_clean.lstrip(os.sep) # No leading separator
+	if local_path_clean.lower().endswith(".png"):
+		local_path_clean = local_path_clean[:-4]
+	texture_path_fragment = local_path_clean + ".png"
+
+	# Retrieval of image file:
+	# --- A: Check Models Origin Resource Pack ---
+	# Search in the resource pack folder where the model JSON file is
+	try:
+		model_path_obj = Path(model_filepath).resolve()
+		pack_root = None
+
+		# Walk up the directory tree to find the 'assets' folder
+		for parent in [model_path_obj] + list(model_path_obj.parents):
+			if (parent / "assets").is_dir():
+				pack_root = parent
+				break
+
+		if pack_root:
+			path_model = pack_root / "assets" / namespace / "textures" / texture_path_fragment
```

Comment:
For this, [`joinpath`](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.joinpath) would be preferable since it's more clear what's going on.